compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+    # also mean that the processing doesn't roundtrip when sem-checking the
+    # result and then passing it back into the sanitizer. Instead, we discard
+    # the check part
+    checkSonsLen(n, 2)
+    if n[0].kind == nkDotExpr:
+      result = parseDotExpr(c, n[0])
+    else:
+      invalid()
+  of nkDerefExpr:
+    checkSonsLen(n, 1)
+    result = newTreeI(nkDerefExpr, n.info, toUntyped n[0])
+  of nkIfExpr, nkIfStmt, nkWhenStmt:
+    # treat all three nodes as an expression when we're in an expression
+    # context
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in n.pairs:
+      result[i] =
+        case it.kind
+        of nkElifExpr, nkElifBranch:
+          parseBranch(c, it, n.kind == nkIfExpr)
+        of nkElseExpr, nkElse:
+          # XXX: hm, also make sure that the 'else' is the last item? Or leave
+          #      that to sem?
+          guardLen(it, 1):
+            newTreeI(it.kind, it.info):
+              if it.kind == nkElseExpr: expr(c, it[0]) else: stmt(c, it[0])
+        else:
+          invalidAst(c, n[i])
+
+  of nkLambda, nkDo:
+    result = parseProcExpr(c, true, n)
+  of nkAccQuoted:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      case n[i].kind
+      of nkIntKinds:
+        result[i] = ast.newIntNode(n[i].kind, n[i].intVal)
+      of nkIdent:
+        result[i] = newIdentNode(n[i].ident, n.info)
+      of nkSym:
+        result[i] = safeSymToIdent(c, n[i])
+      else:
+        result[i] = invalidAst(c, n[i])
+
+  of nkTableConstr:
+    # must have at least one sub node -- it'd be a ``nkCurly`` otherwise
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in n.pairs:
+      result[i] =
+        if it.kind == nkExprColonExpr:
+          guardLen(it, 2):
+            newTreeI(it.kind, it.info, expr(c, it[0]), expr(c, it[1]))
+        else:
+          invalidAst(c, it)
+
+  of nkBind:
+    checkSonsLen(n, 1)
+    result = newTreeI(nkBind, n.info): expr(c, n[0])
+  of nkClosedSymChoice, nkOpenSymChoice:
+    # TODO: revisit
+    result = collapseSymChoice(c, n, true)
+  of nkHiddenStdConv, nkHiddenSubConv:
+    # QUESTION: does validating the nodes make sense if we're skipping them?
+    checkSonsLen(n, 2)
+    result = expr(c, n[1])
+  of nkConv:
+    # ``nkConv`` is not part of the untyped AST, so we translate it into an
+    # ``nkCall``
+    checkSonsLen(n, 2)
+    result = newTreeI(nkCall, n.info): [parseTypeExpr(c, n[0]), expr(c, n[1])]
+  of nkCast:
+    checkSonsLen(n, 2)
+    result = newTreeI(nkCast, n.info): [parseTypeExpr(c, n[0]), expr(c, n[1])]
+  of nkStaticExpr:
+    # skip
+    checkSonsLen(n, 1)
+    result = expr(c, n[0])
+  of nkAddr:
+    checkSonsLen(n, 1)
+    result = newTreeI(nkAddr, n.info): expr(c, n[0])
+  of nkHiddenAddr, nkHiddenDeref:
+    checkSonsLen(n, 1)
+    result = expr(c, n[0])
+  of nkObjDownConv, nkObjUpConv:
+    # XXX: these don't exist in typed AST prior to ``transf``. How should they
```

Comment:
The types might change so the conversions should be deduced again, IMO.

Reply:
Okay, I'm going to check how well the following unwinding works then:
```
ObjDownConv (x) -> Call (type-of-destination) (x) 
```